### ISSUE TITLE
Requirements do not have a motivation for `executable name`

### ISSUE THREAD

**Issue by dgarijo** [SoftwareUnderstanding/software_types#9]
I guess the rationale is to have the name of the package? e.g., in python the name of the executable package you invoke in the command line. It's unclear how would this be in other languages. For example, what a bout a JAR?

**Comment by proycon** [comment]
I added a commit to address this. Does this make it clear?

I think a jar (including the extension) is a valid executableName for java-based applications. Unless the application ships some alias like a shellscript for invocation, then that could be used.

For Java libraries though, the executableName would be the name of the top-level package, as used after ``import``. Which in case of java can be verbose like:

```json
{
   "type": "SoftwareLibrary",
   "name": "Apache Lucene",
   "executableName": "org.apache.lucene"
}
```

For Python libraries, it'd also correspond to the package name, same for most such higher-level languages.

**Comment by dgarijo** [comment]
Hmm, I think I am still a little confused. In the definition you state `It should typically not contain any platform/runtime-specific extensions `, but then in the JAR example you add the JAR extension name. Also, for the library, that's the package name, not the name of anything executable, right?

What if you have a python package but the setup.py is the executable that installs everything? Or you have to run it through "python cli.py" or similar? Would the executable be `cli.py`?

I had understood this property to be the executable command name. What is it exactly that you wish to do with this information? Also, what if you have multiple executables? (e.g., in Widoco you can build different JARs for different target JREs). I am trying to understand better the scope of this property

**Comment by dgarijo** [comment]
At the moment `executableName` is a little overloaded in my opinion

**Comment by proycon** [comment]
> Hmm, I think I am still a little confused. In the definition you state It should typically not contain any platform/runtime-specific extensions , but then in the JAR example you add the JAR extension name. 

Yes, I had updated the documentation a bit but not yet the actual definition:

> Note that the executable name should typically not contain any
> platform/runtime-specific extensions which may differ across systems
> (``.exe``,``.so``,``.dll``,``.dylib``). However, such extensions may be
> included if they are static over all possible systems and needed to invoke the
> software (``.jar``,``.sh``) and a necessary component in invoking the software
> from a specific context.

The idea behind not including executable extensions or library extensions is because those are often too platform-specific. But in case of things like jars, it's pretty static to we can just include it.

> What if you have a python package but the setup.py is the executable that installs everything?

Good point, `executableName` should generally not point to any installer but to the executable *after* installation.

> Or you have to run it through 
> "python cli.py" or similar? Would the executable be cli.py?

The executable would indeed be just `cli.py` (with extension) if that's how you run it. However, ideally python packages with CLI interfaces define entrypoints/console-scripts in their `setup.py`/`pyproject.toml` (see for example: https://github.com/proycon/foliatools/blob/master/setup.py#L35), these translate into tiny wrapper script in some `/usr/bin/` directory that do an import and call a main function. In that case you'll just have something like `cli` (no extension) for `executableName` . Whilst I'm writing this I do suddenly wonder if Python does this differently (with an extension?) on non-unix systems like Windows.

This example ties in with another of your questions:

> What is it exactly that you wish to do with this information?

I want to be able to translate these 'entrypoints' and incorporate them as part of the software metadata. I think they convey some valuable information, which is also useful to present to users. For example, that same software project I mentioned is (automatically) converted to codemeta and presented as follows: https://tools.dev.clariah.nl/foliatools   (ideally the individual tools would have descriptions but I haven't automated that yet, setuptools doesn't provide it either).

> I had understood this property to be the executable command name. What is it exactly that you wish to do with this information? 

Yes, the precise interpretation would differ a bit according to what the run-time context is, the name of the executable command is one of the most common interpretations for most interface types.

> Also, for the library, that's the package name, not the name of anything executable, right?

Depends a bit on the run-time context. 

> At the moment executableName is a little overloaded in my opinion

This is indeed a fair point. Initially I was hoping to keep things simple by just introducing one new property and trying to keep it generic. But if keeping it generic causes too much ambiguity or confusion it might be worth explicitly splitting it.

Let me try work out how we could split it and see if we like that better:

* ``executableName`` - The filename of the executable for this application. If there's an extension (e.g. `exe`) that differs across targeted platforms, the extension should be omitted. This covers only the base filename after installation, not a full path.
* ``schema:executableLibraryName`` - *Library file name e.g., mscorlib.dll, system.web.dll.* If there's an extension component that differs across targeted platforms (e.g. `dll`, `so`, `dylib`), it should be omitted. This covers only the base filename after installation, not a full path.
  * This is an existing schema.org property (used on `APIReference` classes) which would fit our needs and we could just reuse I think. The description part in italics is the existing schema.org definition  (a bit windows-centric unfortunately), the part after that is my addition. In the normal unix world, `executableLibraryName` is something that can be `dlopen()`ed.
* ``importableName`` - the name of a package/module as it is importable in a specific run-time context (via statement like `import`, `use` or `include`). This is used for libraries in languages like Python, Java, R, Perl. Use only the top-level name.
* ``packageName`` - the name of a package as it can be installed by an (unspecified) package manager
* ``containerName`` - the name of the container as it can be run by a container platform, preferably with explicit reference to a container registry (e.g. ``docker.io/proycon/frog:0.25``). The container platform is not strictly defined. 

Would that be better? I'm actually leaning that way too now, although it does result in more properties, it's more precise but also more flexible as we're not constrained to just a single property. I'm a bit in doubt about the last two (for `SoftwarePackage` and `SoftwareImage`) and about how specific we want to be there. I feel we need an additional `packagingSystem` and `containerPlatform` properties to make that useful. 

In line with your earlier comment, we could even add a:

* ``installerName`` - The name of an executable file that installs/builds this application (e.g ``setup.py``)

**Comment by dgarijo** [comment]
@proycon it looks like we are working on similar things (https://software.oeg.fi.upm.es/ is the software catalog I am testing for our organization, built with https://github.com/oeg-upm/soca, and using https://github.com/SoftwareUnderstanding/inspect4py and somef)

I would prefer to have things simple at the moment as well. PackageName, containerName, etc. could be captured with `name`, already present in schema.org.  And maybe we can make the `executableName` more narrow and less overloaded with something like:
```
executableName: (domain: SoftwareApplication) name of the command or executable required to run the target software application within a certain run-time context.  
```
And then we would leave out the part about the extensions, but we could add examples. For instance with `.sh`  and `.jar`, but explaining that `.dll` would not be executableName.

And then containers would not have that property (we can also explain that with an example). 

In short: I propose to have only `executableName`, but simplifying the definition. What do you think?

**Comment by proycon** [comment]
>it looks like we are working on similar things (https://software.oeg.fi.upm.es/ is the software catalog I am testing for our organization, built with https://github.com/oeg-upm/soca, and using https://github.com/SoftwareUnderstanding/inspect4py and somef)

Very interesting! I hadn't seen that yet. It indeed has some overlap with what we are doing at https://github.com/CLARIAH/tool-discovery (deployed at https://tools.dev.clariah.nl/), powered by https://github.com/proycon/codemeta-server,  https://github.com/proycon/codemeta-harvester, and especially https://github.com/proycon/codemetapy .

I'm keeping track of some related initiatives and collaboration possibilities in CLARIAH/clariah-plus#128 . Your project might be worth adding as well.

Back on the main topic:

> I would prefer to have things simple at the moment as well. PackageName, containerName, etc. could be captured with name, already present in schema.org. 

I interpret the `schema:name` on `SoftwareApplication` or one of the subtypes we define as a presentational and human readable name whereas the proposed packageName and containerName would be more identifier. In that sense, perhaps `schema:identifier` might even be appropriate already. In the spirit of keeping things simple, perhaps we should indeed wait with those extra properties and decide on this simpler case first. I don't have an immediate use-case for things beyond `executableName` yet either.

So I agree it's for the best not to conflate our `executableName` with too much and stick to a narrow definition you proposed. 

> executableName: (domain: SoftwareApplication) name of the command or executable required to run the target software application within a certain run-time context. 

I'd rather keep the definition even narrower, it should basically be a filename of an executable file. If we use the word 'command' then it suggests people might also specify parameters etc, but that's out of scope I think. The main goal of `executableName` is simply to identify the particular `targetProduct` unambiguously. It doesn't  provide any further particular instructions on how to properly invoke it. I used the word "certain run-time context" before as we were being more generic, but now we are being more specific we can also lose that and just say:

`executableName: (domain: SoftwareApplication) filename of the executable for the software application`

> And then we would leave out the part about the extensions, but we could add examples. For instance with .sh and .jar, but explaining that .dll would not be executableName.

I'm okay with keeping that out of the definition itself and in a note or examples somewhere yeah. I realize the existing schema.org definitions tend to be fairly broad and brief usually. An important note though is that I think 'executableName` should be a base filename, not a full path, not sure if we should add that to the definition.

I'll do a new PR for this shortly, and then afterwards we can see if we want to incorporate more.

**Comment by dgarijo** [comment]
I reviewed the PR and merged it, I think it looks good now :)

Btw, I was looking at some of the parsers in `codemetapy` and I would like to extend them and add them as part of somef, because it's an alternative way of extracting metadata we plan to support. But the license is incompatible with the one I have in my package (yours is copyleft, mine isn't). Any reason why you have copyleft licenses instead of mit/apache2? I would prefer avoid reimplementing these parsers if possible

**Comment by dgarijo** [comment]
Closing issue, as this seems resolved

**Comment by proycon** [comment]
> Btw, I was looking at some of the parsers in `codemetapy` and I would like to extend them and add them as part of somef, because it's an alternative way of extracting metadata we plan to support. But the license is incompatible with the one I have in my package (yours is copyleft, mine isn't). Any reason why you have copyleft licenses instead of mit/apache2? I would prefer avoid reimplementing these parsers if possible.

Yes, I indeed always use GPL or even AGPL to ensure everything always remains free software (for ideological reasons) and can't be appropriated by any closed-source product. But I don't mind granting you explicit permission hereby to include any parts of codemetapy that I coded and will be coding (that'll be most of it at this stage, check `git blame` to be sure) under MIT or Apache2 in somef and associated projects. I agree that collaboration and avoiding reimplementation is best.

**Comment by dgarijo** [comment]
Awesome, thanks!
